README.md
# A [Cisco](https://www.cisco.com) Wireless LAN Controller plugin for [Homebridge](https://github.com/nfarina/homebridge).  

This creates a Light Switch in homekit for the status of an SSID using RESTCONF.

![alt text](example.png "Dashboard Example Screenshot")

This was tested to work with the Cisco Catalyst 9800CL. It should work with any other model of Cisco Catalyst 9800.

# Installation
Run these commands:

    % sudo npm install -g homebridge
    % sudo npm install -g homebridge-c9800


NOTE: If you install homebridge like this:

    sudo npm install -g --unsafe-perm homebridge

Then all subsequent installations must be like this:

    sudo npm install -g --unsafe-perm homebridge-c9800

# Configuration
Example accessory config (needs to be added to the homebridge config.json):
 ...

		"accessories": [
			{
				"name": "Guest WiFi",
				"ipAddress": "192.168.1.115",
				"username": "admin",
				"password": "password",
				"wlanName": "guest",
				"model": "C9800CL",
				"serial": "myserialno",
				"timeout": 10000,
				"debug": false,
				"accessory": "Cisco 9800"
			}
		]
 ...

### Config Explanation:

Field           						| Description
----------------------------|------------
**accessory**   						| (required) Must always be "Cisco 9800".
**name**										| (required) The name you want to use for the light switch widget.
**ipAddress**								| (required) The IP address of the WLC (should be static, not DHCP).
**username**								| (required) The username used to access the WLC.
**password**								| (required) The password used to access the WLC.
**wlanName**								| (required) The WLAN Name associated with the SSID on the WLC.
**model**										| (optional) This shows up in the homekit accessory Characteristics.
**serial**									| (optional) This shows up in the homekit accessory Characteristics.
**timeout**									| (optional) The timeout duration in ms for the web API calls.
**debug**										| (optional) Enables additional logging.

To make your WLC work with the plugin:

1. Connect your WLC to your network.
2. Configure your WLC.
3. Enable netconf on the WLC.
4. Enable restconf on the WLC.
5. Write down the IP address of the WLC.
6. Create your config file according to the above example (or using the Homebridge UI).


# More information
Check out https://github.com/nfarina/homebridge for more information about Homebridge.

## License
This project is licensed to you under the terms of the [Cisco Sample
Code License](./LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 0.0.1 (2021-10-05)

#### Changes

- Initial commit.


index.js
/*

Copyright (c) 2021 Cisco and/or its affiliates.

This software is licensed to you under the terms of the Cisco Sample
Code License, Version 1.0 (the "License"). You may obtain a copy of the
License at

               https://developer.cisco.com/docs/licenses

All use of the material herein must be in accordance with the terms of
the License. All rights not expressly granted by the License are
reserved. Unless required by applicable law or agreed to separately in
writing, software distributed under the License is distributed on an "AS
IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
or implied.

*/

const axios = require('axios');
axios.defaults.withCredentials = true;

const https = require('https');

const query_path = "/restconf/data/Cisco-IOS-XE-wireless-wlan-cfg:wlan-cfg-data/wlan-cfg-entries/wlan-cfg-entry=";

const timeout = 10000;

const debug = false;

var Service;
var Characteristic;

module.exports = function (homebridge) {
    Service = homebridge.hap.Service;
    Characteristic = homebridge.hap.Characteristic;
    homebridge.registerAccessory("homebridge-c9800", "Cisco 9800", C9800);
};


function C9800(log, config) {
    this.log = log;
    this.ipAddress = config["ipAddress"];
    this.username = config["username"];
    this.password = config["password"];
    this.webConfig = {
			data: {},
			auth: {
				username: this.username,
				password: this.password
			},
			headers: {
				Accept: "application/yang-data+json",
				"Content-Type": "application/yang-data+json"
			},
		};
    this.wlan = config["wlanName"];
    this.model = config["model"] === undefined ? "" : config["model"];
    this.serial = config["serial"] === undefined ? "" : config["serial"];
    this.name = config["name"];
    this.timeout = config["timeout"] === undefined ? timeout : config["timeout"];
    this.debug = config["debug"] === undefined ? debug : config["debug"];
		this.api = axios.create({
			timeout: this.timeout,
			httpsAgent: new https.Agent({  
    		rejectUnauthorized: false
		  }),
		});

}

C9800.prototype = {

  	getPowerState: function (callback) {
       
				this.api.get('https://' + this.ipAddress + query_path + this.wlan, this.webConfig)
					.then(resp => {
					  if (this.debug) {
							this.log("https://" + this.ipAddress + query_path + this.wlan);
							this.log("WLC response: " + resp["data"]["Cisco-IOS-XE-wireless-wlan-cfg:wlan-cfg-entry"]["apf-vap-id-data"]["wlan-status"]);
        		}

						callback(null, resp["data"]["Cisco-IOS-XE-wireless-wlan-cfg:wlan-cfg-entry"]["apf-vap-id-data"]["wlan-status"] === true);
					})
					.catch(err => {
						callback(err)
					});
    },

    setPowerState: function(powerOn, callback) {
				let newData;
				this.api.get('https://' + this.ipAddress + query_path + this.wlan, this.webConfig)
					.then(resp => {
						newData = {"Cisco-IOS-XE-wireless-wlan-cfg:wlan-status": powerOn};
					  if (this.debug) {
							this.log("https://" + this.ipAddress + query_path + this.wlan);
							this.log("WLC response: " + resp["data"]["Cisco-IOS-XE-wireless-wlan-cfg:wlan-cfg-entry"]["apf-vap-id-data"]["wlan-status"]);
							this.log (newData)
						}
						this.api.patch('https://' + this.ipAddress + query_path + this.wlan + "/apf-vap-id-data/wlan-status", newData, this.webConfig)
							.then(resp => {
							  if (this.debug) {						
									this.log("Patched SSID on WLC to " + powerOn);
								}
							})
						callback();
					})
					.catch(err => {
						callback(err)
					})
					
    },

    getServices: function () {
        const informationService = new Service.AccessoryInformation();

        informationService
            .setCharacteristic(Characteristic.Manufacturer, "Cisco")
            .setCharacteristic(Characteristic.Model, this.model)
            .setCharacteristic(Characteristic.SerialNumber, this.serial);

        switchService = new Service.Switch(this.name);
        switchService
            .getCharacteristic(Characteristic.On)
                .on('get', this.getPowerState.bind(this))
                .on('set', this.setPowerState.bind(this));

        this.informationService = informationService;
        this.switchService = switchService;
        return [informationService, switchService];
    }
};
